### PR TITLE
Add meta-information to esil-ops ##esil

### PR DESCRIPTION
**Pull request by condret**
Adding meta-info so that one could perform analysis based on these, which includes custom-ops ... type-field still needs to be done

### CODE REVIEW

**Comment by radare** [comment]
pls fix the cornflakes

File: libr/include/r_anal.h [hunk 1]
```diff
@@ -1178,7 +1179,28 @@ typedef struct r_anal_esil_t {
...
+	R_ANAL_ESIL_OP_TYPE_MATH = 0x10,
+	R_ANAL_ESIL_OP_TYPE_CUSTOM = 0x20
+};
+#endif
+
+typedef bool (*RAnalEsilOpCb)(RAnalEsil *esil);
+
+typedef struct r_anal_esil_operation_t {
+	RAnalEsilOpCb code;
+	ut32 push;
```
Comment (lowlyw) [inline libr/include/r_anal.h]: add comment, what does push and pop here mean...

File: libr/core/esil_data_flow.c [hunk 2]
```diff
@@ -387,13 +302,13 @@ static int edf_use_new_push_1 (RAnalEsil *esil, const char *op_string, AddConstr
 	return r_anal_esil_push (esil, r_strbuf_get(result->content));
 }
 
-static int edf_consume_1_use_old_new_push_1 (RAnalEsil *esil, const char *op_string, AddConstraintStringConsume1UseOldNewCB cb) {
+static bool edf_consume_1_use_old_new_push_1 (RAnalEsil *esil, const char *op_string, AddConstraintStringConsume1UseOldNewCB cb) {
 	EsilDataFlow *edf = (EsilDataFlow *)esil->user;
 	char *src = r_anal_esil_pop(esil);
 
 	if (!src) {
 		free (src);
```
Comment (lowlyw) [inline libr/core/esil_data_flow.c]: rm this free, `if (!x) free(x)` is a useless pattern

File: libr/core/esil_data_flow.c [hunk 3]
```diff
@@ -313,16 +227,17 @@ static int edf_consume_2_push_1 (RAnalEsil *esil, const char *op_string) {
...
-static int edf_consume_1_push_1(RAnalEsil *esil, const char *op_string) {
+static bool edf_consume_1_push_1(RAnalEsil *esil) {
+	const char *op_string = esil->current_opstr;
 	EsilDataFlow *edf = (EsilDataFlow *)esil->user;
 	char *src = r_anal_esil_pop(esil);
 
 	if (!src) {
 		free (src);
-		return 0;
+		return false;
```
Comment (lowlyw) [inline libr/core/esil_data_flow.c]: rm the free of src, `if (!x) free(x)` is useless

**Comment by radare** [comment]
still conflicts. please rebase

File: libr/anal/esil.c [hunk 4]
```diff
@@ -105,17 +105,30 @@ R_API RAnalEsil *r_anal_esil_new(int stacksize, int iotrap, unsigned int addrsiz
...
 	char t[128];
 	if (!code || !op || !strlen (op) || !esil || !esil->ops) {
 		return false;
 	}
 	char *h = sdb_itoa (sdb_hash (op), t, 16);
-	sdb_num_set (esil->ops, h, (ut64)(size_t)code, 0);
-	if (!sdb_num_exists (esil->ops, h)) {
-		eprintf ("can't set esil-op %s\n", op);
-		return false;
+	RAnalEsilOp *operation = sdb_num_get (esil->ops, h, 0);
```
Comment (radare) [inline libr/anal/esil.c]: uh? sdb_num_get returns ut64, not void * that should at least warn without any sizet cast

File: libr/anal/esil.c [hunk 5]
```diff
@@ -105,17 +105,30 @@ R_API RAnalEsil *r_anal_esil_new(int stacksize, int iotrap, unsigned int addrsiz
...
+	if (!operation) {
+		operation = R_NEW (RAnalEsilOpCb);
+		if (!operation) {
+			eprintf ("cannot allocate esil-operation %s\n", op);
```
Comment (radare) [inline libr/anal/esil.c]: s/cannot/Cannot

File: libr/anal/esil.c (hunk 4 above)
Comment (radare) [reply libr/anal/esil.c]: can we use "op" instead of "operation" ?

File: libr/anal/esil.c [hunk 6]
```diff
@@ -105,17 +105,30 @@ R_API RAnalEsil *r_anal_esil_new(int stacksize, int iotrap, unsigned int addrsiz
...
+			return false;
+		}
+		sdb_num_set (esil->ops, h, (ut64)(size_t)operation, 0);
+		if (!sdb_num_exists (esil->ops, h)) {
+			eprintf ("cannot set esil-operation %s\n", op);
```
Comment (radare) [inline libr/anal/esil.c]: s/cannot/Cannot

File: libr/anal/esil.c [hunk 7]
```diff
@@ -150,13 +163,20 @@ R_API bool r_anal_esil_set_pc(RAnalEsil *esil, ut64 addr) {
 	return false;
 }
 
+static int esil_ops_free_cb(void *user, const char *k, const char *v) {
+	RAnalEsilOp *op = (RAnalEsilOp *)(size_t)sdb_atoi (v);
```
Comment (radare) [inline libr/anal/esil.c]: wat, this is unused code, just return true and drop this line

File: libr/core/esil_data_flow.c [hunk 8]
```diff
@@ -254,17 +167,18 @@ static int edf_consume_2_set_reg (RAnalEsil *esil, const char *op_string) {
...
 	free (dst);
-	return 1;
+	return true;
 }
 
-static int edf_consume_2_push_1 (RAnalEsil *esil, const char *op_string) {
+static bool edf_consume_2_push_1 (RAnalEsil *esil) {
+	const char *op_string = esil->current_opstr;
 	EsilDataFlow *edf = (EsilDataFlow *)esil->user;
 	char *src[2] = {r_anal_esil_pop(esil), r_anal_esil_pop(esil)};
```
Comment (radare) [inline libr/core/esil_data_flow.c]: miossing space before (

File: libr/core/esil_data_flow.c [hunk 9]
```diff
@@ -472,44 +379,41 @@ R_API void r_core_anal_esil_graph(RCore *core, const char *expr) {
...
-	r_anal_esil_set_op(esil, "<<", edf_shl);
-	r_anal_esil_set_op(esil, ">>>", edf_rotr);
-	r_anal_esil_set_op(esil, ">>>", edf_rotl);
-	r_anal_esil_set_op(esil, "!", edf_not);
-	r_anal_esil_set_op(esil, "[1]", edf_peek_1);
-	r_anal_esil_set_op(esil, "[2]", edf_peek_2);
-	r_anal_esil_set_op(esil, "[4]", edf_peek_4);
-	r_anal_esil_set_op(esil, "[8]", edf_peek_8);
-	r_anal_esil_set_op(esil, "[16]", edf_peek_16);
+	r_anal_esil_set_op(esil, "=", edf_consume_2_set_reg, 0, 2, R_ANAL_ESIL_OP_TYPE_REG_WRITE);
```
Comment (radare) [inline libr/core/esil_data_flow.c]: missing space before ( in all those lines

File: libr/anal/esil.c [hunk 10]
```diff
@@ -105,17 +105,30 @@ R_API RAnalEsil *r_anal_esil_new(int stacksize, int iotrap, unsigned int addrsiz
 	return esil;
 }
 
-R_API int r_anal_esil_set_op(RAnalEsil *esil, const char *op, RAnalEsilOp code) {
+R_API bool r_anal_esil_set_op(RAnalEsil *esil, const char *op, RAnalEsilOpCb code, ut32 push, ut32 pop, ut32 type) {
 	char t[128];
 	if (!code || !op || !strlen (op) || !esil || !esil->ops) {
```
Comment (radare) [inline libr/anal/esil.c]: use r_return_val_if_fail

File: libr/anal/esil.c (hunk 4 above)
Comment (radare) [reply libr/anal/esil.c]: ping, fix this

File: libr/anal/esil.c (hunk 6 above)
Comment (radare) [reply libr/anal/esil.c]: use uppercase

**Comment by radare** [comment]
1 week left for the release. can you pls finish this pr? please rebase it and address the comments to make travis green

**Comment by radare** [comment]
PING

File: libr/anal/esil.c (hunk 7 above)
Comment (condret) [reply libr/anal/esil.c]: this code is used, and passed to sdb_foreach in r_anal_esil_free

File: libr/anal/esil.c (hunk 4 above)
Comment (condret) [reply libr/anal/esil.c]: op is already used as an argument passed to this function

Comment (condret) [reply libr/anal/esil.c]: i will use eop

**Comment by radare** [comment]
Its still pretty red

File: libr/anal/esil.c [hunk 11]
```diff
@@ -105,17 +105,28 @@ R_API RAnalEsil *r_anal_esil_new(int stacksize, int iotrap, unsigned int addrsiz
 	return esil;
 }
 
-R_API int r_anal_esil_set_op(RAnalEsil *esil, const char *op, RAnalEsilOp code) {
+R_API bool r_anal_esil_set_op(RAnalEsil *esil, const char *op, RAnalEsilOpCb code, ut32 push, ut32 pop, ut32 type) {
+	r_return_val_if_fail(code && op && strlen (op) && esil && esil->ops, false);
```
Comment (radare) [inline libr/anal/esil.c]: Missing space

Comment (radare) [reply libr/anal/esil.c]: May be the reason why travis is failing. Run the tests in local and see if now is failing bwcause of that assert

**Comment by radare** [comment]
kill or rebase your r2r branch

File: libr/anal/esil.c [hunk 12]
```diff
@@ -105,17 +105,28 @@ R_API RAnalEsil *r_anal_esil_new(int stacksize, int iotrap, unsigned int addrsiz
...
-		eprintf ("can't set esil-op %s\n", op);
-		return false;
+	RAnalEsilOp *eop = (RAnalEsilOp *)(size_t)sdb_num_get (esil->ops, h, 0);
+	if (!eop) {
+		eop = R_NEW (RAnalEsilOpCb);
+		if (!eop) {
+			eprintf ("Cannot allocate esil-operation %s\n", op);
+			return false;
+		}
+		sdb_num_set (esil->ops, h, (ut64)(size_t)eop, 0);
```
Comment (radare) [inline libr/anal/esil.c]: crash happens inside sdb_num_set because of heap corruption

File: libr/core/cmd_search.c [hunk 13]
```diff
@@ -1625,7 +1625,7 @@ static void do_esil_search(RCore *core, struct search_parameters *param, const c
 		}
 		/* hook addrinfo */
 		core->anal->esil->cb.user = core;
-		r_anal_esil_set_op (core->anal->esil, "AddrInfo", esil_addrinfo);
+		r_anal_esil_set_op (core->anal->esil, "AddrInfo", esil_addrinfo, 1, 1, R_ANAL_ESIL_OP_TYPE_UNKNOWN);
```
Comment (radare) [inline libr/core/cmd_search.c]: segfault in just a step
IMAGE 2019-06-20 12:15:25

File: libr/anal/esil.c [hunk 14]
```diff
@@ -3028,149 +3048,157 @@ R_API int r_anal_esil_condition(RAnalEsil *esil, const char *str) {
...
+	OP ("+=[]", esil_mem_addeq, 0, 2, OT_MATH | OT_MEMR | OT_MEMW);
+	OP ("+=[1]", esil_mem_addeq1, 0, 2, OT_MATH | OT_MEMR | OT_MEMW);
+	OP ("+=[2]", esil_mem_addeq2, 0, 2, OT_MATH | OT_MEMR | OT_MEMW);
+	OP ("+=[4]", esil_mem_addeq4, 0, 2, OT_MATH | OT_MEMR | OT_MEMW);
+	OP ("+=[8]", esil_mem_addeq8, 0, 2, OT_MATH | OT_MEMR | OT_MEMW);
+	OP ("-=[]", esil_mem_subeq, 0, 2, OT_MATH | OT_MEMR | OT_MEMW);
+	OP ("-=[1]", esil_mem_subeq1, 0, 2, OT_MATH | OT_MEMR | OT_MEMW);
+	OP ("-=[2]", esil_mem_subeq2, 0, 2, OT_MATH | OT_MEMR | OT_MEMW);
+	OP ("-=[4]", esil_mem_subeq4, 0, 2, OT_MATH | OT_MEMR | OT_MEMW);
+	OP ("-=[8]", esil_mem_subeq8, 0, 2, OT_MATH | OT_MEMR | OT_MEMW);
```
Comment (radare) [inline libr/anal/esil.c]: zillions of warnings
IMAGE 2019-06-20 12:15:05

**Comment by radare** [comment]
i fixed the warnings and its now green but i cant commit to your branch. so pls apply this patch:

```
$ git diff |base64
ZGlmZiAtLWdpdCBhL2xpYnIvYW5hbC9lc2lsLmMgYi9saWJyL2FuYWwvZXNpbC5jCmluZGV4IDU0M2FkYzFhMS4uZjM4MzMzNDAwIDEwMDY0NAotLS0gYS9saWJyL2FuYWwvZXNpbC5jCisrKyBiL2xpYnIvYW5hbC9lc2lsLmMKQEAgLTExMSw3ICsxMTEsNyBAQCBSX0FQSSBib29sIHJfYW5hbF9lc2lsX3NldF9vcChSQW5hbEVzaWwgKmVzaWwsIGNvbnN0IGNoYXIgKm9wLCBSQW5hbEVzaWxPcENiIGNvZAogCWNoYXIgKmggPSBzZGJfaXRvYSAoc2RiX2hhc2ggKG9wKSwgdCwgMTYpOwogCVJBbmFsRXNpbE9wICplb3AgPSAoUkFuYWxFc2lsT3AgKikoc2l6ZV90KXNkYl9udW1fZ2V0IChlc2lsLT5vcHMsIGgsIDApOwogCWlmICghZW9wKSB7Ci0JCWVvcCA9IFJfTkVXIChSQW5hbEVzaWxPcENiKTsKKwkJZW9wID0gUl9ORVcwIChSQW5hbEVzaWxPcCk7CiAJCWlmICghZW9wKSB7CiAJCQllcHJpbnRmICgiQ2Fubm90IGFsbG9jYXRlIGVzaWwtb3BlcmF0aW9uICVzXG4iLCBvcCk7CiAJCQlyZXR1cm4gZmFsc2U7CkBAIC0xODA4LDcgKzE4MDgsNiBAQCBzdGF0aWMgYm9vbCBlc2lsX3Bva2UoUkFuYWxFc2lsICplc2lsKSB7CiB9CiAKIHN0YXRpYyBib29sIGVzaWxfcG9rZV9zb21lKFJBbmFsRXNpbCAqZXNpbCkgewotCWJvb2wgcmV0ID0gZmFsc2U7CiAJaW50IGksIHJlZ3NpemU7CiAJdXQ2NCBwdHIsIHJlZ3MgPSAwLCB0bXA7CiAJY2hhciAqY291bnQsICpkc3QgPSByX2FuYWxfZXNpbF9wb3AgKGVzaWwpOwpAQCAtMTgzNCw4ICsxODMzLDggQEAgc3RhdGljIGJvb2wgZXNpbF9wb2tlX3NvbWUoUkFuYWxFc2lsICplc2lsKSB7CiAJCQkJCS8qIFRPRE86IGltcGxlbWVudCBwZWVrIGhlcmUgKi8KIAkJCQkJLy8gcmVhZCBmcm9tICRkc3QKIAkJCQkJcl93cml0ZV9ibGUgKGIsIG51bTY0LCBlc2lsLT5hbmFsLT5iaWdfZW5kaWFuLCByZWdzaXplKTsKLQkJCQkJcmV0ID0gcl9hbmFsX2VzaWxfbWVtX3dyaXRlIChlc2lsLCBwdHIsIGIsIEJZVEVTX1NJWkUpOwotCQkJCQlpZiAocmV0ICE9IEJZVEVTX1NJWkUpIHsKKwkJCQkJaW50IHJvYWQgPSByX2FuYWxfZXNpbF9tZW1fd3JpdGUgKGVzaWwsIHB0ciwgYiwgQllURVNfU0laRSk7CisJCQkJCWlmIChyb2FkICE9IEJZVEVTX1NJWkUpIHsKIAkJCQkJCS8vZXByaW50ZiAoIkNhbm5vdCB3cml0ZSBhdCAweCUwOCIgUEZNVDY0eCAiXG4iLCBwdHIpOwogCQkJCQkJZXNpbC0+dHJhcCA9IDE7CiAJCQkJCX0KQEAgLTE4NDUsMTEgKzE4NDQsMTEgQEAgc3RhdGljIGJvb2wgZXNpbF9wb2tlX3NvbWUoUkFuYWxFc2lsICplc2lsKSB7CiAJCQl9CiAJCQlmcmVlIChkc3QpOwogCQkJZnJlZSAoY291bnQpOwotCQkJcmV0dXJuIDE7CisJCQlyZXR1cm4gdHJ1ZTsKIAkJfQogCQlmcmVlIChkc3QpOwogCX0KLQlyZXR1cm4gMDsKKwlyZXR1cm4gZmFsc2U7CiB9CiAKIC8qIFBFRUsgKi8KQEAgLTE5MjYsNyArMTkyNSw2IEBAIHN0YXRpYyBib29sIGVzaWxfcGVlayhSQW5hbEVzaWwgKmVzaWwpIHsKIH07CiAKIHN0YXRpYyBib29sIGVzaWxfcGVla19zb21lKFJBbmFsRXNpbCAqZXNpbCkgewotCWJvb2wgcmV0ID0gZmFsc2U7CiAJaW50IGk7CiAJdXQ2NCBwdHIsIHJlZ3M7CiAJLy8gcG9wIHB0cgpAQCAtMTk0Niw4ICsxOTQ0LDggQEAgc3RhdGljIGJvb2wgZXNpbF9wZWVrX3NvbWUoUkFuYWxFc2lsICplc2lsKSB7CiAJCQkJCQlFUlIgKCJDYW5ub3QgcG9wIGluIHBlZWsiKTsKIAkJCQkJCXJldHVybiAwOwogCQkJCQl9Ci0JCQkJCXJldCA9IHJfYW5hbF9lc2lsX21lbV9yZWFkIChlc2lsLCBwdHIsIGEsIDQpOwotCQkJCQlpZiAocmV0ID09IHNpemVvZiAodXQzMikpIHsKKwkJCQkJaW50IHJvYWQgPSByX2FuYWxfZXNpbF9tZW1fcmVhZCAoZXNpbCwgcHRyLCBhLCA0KTsKKwkJCQkJaWYgKHJvYWQgPT0gc2l6ZW9mICh1dDMyKSkgewogCQkJCQkJbnVtMzIgPSByX3JlYWRfYmxlMzIgKGEsIGVzaWwtPmFuYWwtPmJpZ19lbmRpYW4pOwogCQkJCQkJcl9hbmFsX2VzaWxfcmVnX3dyaXRlIChlc2lsLCBmb28sIG51bTMyKTsKIAkJCQkJfSBlbHNlIHsKQEAgLTE5NjEsMTEgKzE5NTksMTEgQEAgc3RhdGljIGJvb2wgZXNpbF9wZWVrX3NvbWUoUkFuYWxFc2lsICplc2lsKSB7CiAJCQl9CiAJCQlmcmVlIChkc3QpOwogCQkJZnJlZSAoY291bnQpOwotCQkJcmV0dXJuIDE7CisJCQlyZXR1cm4gdHJ1ZTsKIAkJfQogCQlmcmVlIChkc3QpOwogCX0KLQlyZXR1cm4gMDsKKwlyZXR1cm4gZmFsc2U7CiB9CiAKIC8qIE9SRVEgKi8KQEAgLTI4MzcsNyArMjgzNSw4IEBAIHN0YXRpYyBpbnQgcnVud29yZChSQW5hbEVzaWwgKmVzaWwsIGNvbnN0IGNoYXIgKndvcmQpIHsKIAkJCQkJcmV0dXJuIDE7IC8vIFhYWCBjYW5ub3QgcmV0dXJuICE9IDEKIAkJCQl9CiAJCQl9Ci0JCQllc2lsLT5jdXJyZW50X29wc3RyID0gd29yZDsKKwkJCWZyZWUgKGVzaWwtPmN1cnJlbnRfb3BzdHIpOworCQkJZXNpbC0+Y3VycmVudF9vcHN0ciA9IHN0cmR1cCAod29yZCk7CiAJCQljb25zdCBib29sIHJldCA9IG9wLT5jb2RlIChlc2lsKTsKIAkJCWVzaWwtPmN1cnJlbnRfb3BzdHIgPSBOVUxMOwogCQkJaWYgKCFyZXQpIHsK
```

**Comment by radare** [comment]
please apply my patch

**Comment by condret** [comment]
I took some of your patch, there was a memleak in it, which I didn't adopt, since it was a memleak. Also expecting r_anal_esil_mem_write/read returning int is somehow questionable, if we look at rio, which returns bool for good reason

**Comment by radare** [comment]
I have reopened because i guess you close it by mistake

**Comment by radare** [comment]
confflicts

**Comment by radare** [comment]
lol you commited this into master instead of merging the branch. closing again

**Pull request closed without merging**